api/builtins_qlik/HelmChart.go

```diff
@@ -27,349 +39,385 @@ type HelmChartPlugin struct {
-				return nil, err
-			}
-			if _, err = os.Stat(p.ChartHome); os.IsNotExist(err) {
-				err = p.fetchHelm()
-				if err != nil {
-					p.logger.Printf("error executing fetchHelm(), error: %v\n", err)
-					return nil, err
-				}
-			} else if err != nil {
-				p.logger.Printf("error executing stat on file: %v, error: %v\n", p.ChartHome, err)
-			}
+	return nil
+}
 
-			err = p.deleteRequirements(p.ChartHome)
-			if err != nil {
-				p.logger.Printf("error executing deleteRequirements() for dir: %v, error: %v\n", p.ChartHome, err)
-				return nil, err
-			}
+func (p *HelmChartPlugin) Generate() (resmap.ResMap, error) {
+
+	var templatedYaml []byte
 
-			templatedYaml, err = p.templateHelm()
-			if err != nil {
-				p.logger.Printf("error executing templateHelm(), error: %v\n", err)
+	hashFilePath := filepath.Join(p.hashFolder, p.hash)
+	if _, err := os.Stat(hashFilePath); err != nil {
+		if os.IsNotExist(err) {
+			if templatedYaml, err = p.executeHelmTemplate(); err != nil {
+				p.logger.Printf("error executing helm template, error: %v\n", err)
 				return nil, err
-			}
-			os.MkdirAll(hashfolder, os.ModePerm)
-			if err = ioutil.WriteFile(hashfile, templatedYaml, 0644); err != nil {
-				p.logger.Printf("error writing kustomization yaml to file: %v, error: %v\n", hashfile, err)
+			} else if err = ioutil.WriteFile(hashFilePath, templatedYaml, 0644); err != nil {
+				p.logger.Printf("error writing kustomization yaml to file: %v, error: %v\n", hashFilePath, err)
 			}
 		} else {
 			return nil, err
 		}
 	} else {
-		templatedYaml, err = ioutil.ReadFile(hashfile)
-		if err != nil {
-			p.logger.Printf("error reading file: %v, error: %v\n", hashfile, err)
-			return nil, err
-		}
-	}
-
-	if len(p.ChartPatches) > 0 {
-		templatedYaml, err = p.applyPatches(templatedYaml)
+		templatedYaml, err = ioutil.ReadFile(hashFilePath)
 		if err != nil {
-			p.logger.Printf("error executing applyPatches(), error: %v\n", err)
+			p.logger.Printf("error reading file: %v, error: %v\n", hashFilePath, err)
 			return nil, err
 		}
 	}
 
 	return p.rf.NewResMapFromBytes(templatedYaml)
 }
 
-func (p *HelmChartPlugin) deleteRequirements(dir string) error {
+func (p *HelmChartPlugin) executeHelmTemplate() ([]byte, error) {
+	const repoName = "qlik"
+
+	os.Setenv("HELM_NAMESPACE", p.ReleaseNamespace)
+	os.Setenv("XDG_CONFIG_HOME", p.HelmHome)
+	os.Setenv("XDG_CACHE_HOME", p.HelmHome)
+	settings := cli.New()
+
+	if err := p.helmRepoAdd(settings, repoName, p.ChartRepo); err != nil {
+		p.logger.Printf("error adding repo: %v, err: %v\n", p.ChartRepo, err)
+		return nil, err
+	}
+
+	if err := p.helmReposUpdate(settings); err != nil {
+		p.logger.Printf("error updating helm repos, err: %v\n", err)
+		return nil, err
+	}
+
+	if err := p.helmFetchIfRequired(settings, repoName); err != nil {
+		p.logger.Printf("error fetching from helm, err: %v\n", err)
+		return nil, err
+	}
+
+	subChartPath := filepath.Join(p.ChartHome, p.ChartName, "charts", p.SubChart)
```

Comment:
are we assuming all charts have sub-charts?

Reply:
Good point too! Addressing.

Reply:
@bearium: Please take a look at the updated code when you get a chance.